artemis-selector/src/main/java/org/apache/activemq/artemis/selector/impl/SelectorParser.java

```diff
@@ -80,11 +78,15 @@ public static BooleanExpression parse(String sql) throws FilterException {
                StrictParser parser = new StrictParser(new StringReader(actual));
                e = parser.JmsSelector();
             }
-            cache.put(sql, e);
+            synchronized (cache) {
```

Comment:
Must be a better way than sync blocks.

@franz1981 ideas?

Reply:
I imagine there might be a "better" option here in terms of performance than using `synchronized`, but then again this isn't exactly on the hot path.  This code is only executed when a selector/filter is first created (e.g. for a JMS consumer, for a bridge, etc.), and it is only executed once (i.e. execution isn't looped).  Most situations where cache growth is problematic (e.g. lots of consumers with lots of different selectors connecting and disconnecting often) are generally considered messaging anti-patterns anyway.  For such pathological cases using `synchronized` is potentially quite a bit faster than a complete lack of thread safety because it prevents unfettered cache growth.  In my opinion @franz1981's time is better spent elsewhere as this smacks of premature optimization.

Reply:
@jbertram @michaelandrepearce If it is not an hot path I thing that what Justin is suggesting makes totally sense: anyway if we can afford to put guava in place I believe that https://google.github.io/guava/releases/16.0/api/docs/com/google/common/cache/CacheBuilder.html#maximumSize(long) would be a good option too.
It will scale with a factor specified by the concurrent level specified, but it could (I do not remember TBH) introduce some behavioural changes from the current impl.
I understand both @michaelandrepearce and @jbertram opinions so I think that the general rule to balance changes vs effectiveness on common case apply here: I leave you decide how much this tradeoff apply here...
I'm honored to be "summoned" :+1:

Reply:
FWIW, the LRUCache is also used in `org.apache.activemq.artemis.core.protocol.openwire.OpenWireProtocolManager` and access is controlled with `synchronized` blocks.

Reply:
On that then, i would simply remove the use in the SelectorParser for this particular case.

Re the actual class, if its used in OpenWireProtocolManager then it would suggest the LRUCache is being used beyond its original scope and should move out of its current location into common util, if it is to be re-used in other area's and then if it is to slowly be used in other areas such us the protocol manager, it may warrant some future efforts to harden it.